### PR TITLE
Add option for forcing plugin choice when opening files in the GUI

### LINKED ISSUE #4388: Shift-drag or Alt-drag should always open plugin selection dialog
## 🚀 Feature

With #4347 (and maybe previous PRs), whenever multiple plugins can open a particular file, we get a plugin selection dialog when dragging and dropping a file onto the viewer. This dialog allows users to save the file association to preferences. At that point, if the user wants to use a different plugin under some circumstances, they have to wipe the preference. (I think?) It would be good if shift-drag or alt-drag would *always* open the plugin selection dialog, regardless of preferences.

Additionally, this option could correspond to a menu item, e.g. File > Open with Plugin...

CC @DragaDoncila

### PR DESCRIPTION
**Pull request by DragaDoncila**
# Description
As requested in #4388 and #4613, this PR adds:

- `Alt+Drag` or `Alt+Shift+Drag` for opening files/folder/stacks with a forced plugin choice via the dialog (ignoring preferences)
- A new `File>Open with Plugin...` submenu that supports the three different file opening options with a forced plugin choice via the dialog

I think the new submenu is terrible tbh, but I also think it's the only sane way to do it given the current opening choices. Long term I want to move to just one `File > Open` dialog, where the user can make all their selections, including the files, folders, whether to stack or not and the plugin to use for opening. I've purposely not added shortcuts to the file menu options because I feel like that's inviting chaos but open to ideas.

If we don't want to keep this submenu we can also just keep the force option via drag & drop for now.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Closes #4388

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Review by tlambert03 — APPROVED** [review]
I totally agree that the submenu points to some unification needed in a future PR, but I think this PR is a strict improvement on what we currently have.  Thanks!

as discussed offline, the one additional thing we could consider here is: when someone alt-opens a file, we pre-select the "current" preference in the radio boxes  *or* we uncheck the box to remember the preference

**Comment by DragaDoncila** [comment]
Thanks @tlambert03 I've updated to uncheck the remember checkbox by default. That makes more sense to me because I see this as a temporary preference override. Will leave this open for @jni to weigh in too since he first requested it.

**Comment by sofroniewn** [comment]
I'm going to give this a merge tomorrow morning unless hearing otherwise as I think it's good to go. thanks!!

**Comment by DragaDoncila** [comment]
@sofroniewn just found a bug here in the process of reviewing something else, will fix today and ping when it's ready again. Sorry!

**Comment by DragaDoncila** [comment]
It's fixed now! Might be good if you sanity check this by pulling it down before merge @sofroniewn

File: napari/_qt/dialogs/qt_reader_dialog.py
Comment (Czaki) [inline napari/_qt/dialogs/qt_reader_dialog.py]: Why optional? The default value is bool. 
```suggestion
    plugin_override: bool = False,
```

File: napari/_qt/dialogs/qt_reader_dialog.py
Comment (Czaki) [inline napari/_qt/dialogs/qt_reader_dialog.py]:
```suggestion
    plugin_override: bool
```

File: napari/_qt/menus/file_menu.py
Comment (Czaki) [inline napari/_qt/menus/file_menu.py]: I do not like the `lambda` here. It may lead to leaking memory. 
`psygnal` handle `partial` properly but ono other of our used signal system handle it properly. And If I correctly remember, non of them handle properly `lambda`


I prefer to create a private method for this instead of `lambda`. Also this may be an inspiration for new contributors to use `lambda` in more fragile places

**Review by Czaki — COMMENTED** [review]
Using Alt for enforcing the selection plugin is a great feature, but it needs to be documented to increase discoverability.

**Comment by VolkerH** [comment]
Just late to this PR here and maybe this is not the right place to comment.
But just thinking about other software packages and their user interfaces, there is often a
`File->Open`, `File->Save` for the software packages native file formats and a `File->Import` and `File->Export` (or `File->Save As`) for non-native file formats.
In a similar fashion built-in readers/writers could be under `File->Open`, `File->Save` and file formats supported via plugins under  `File->Import` and `File->Export`.  That may be slightly more preferable over `File>Open with Plugin...`

**Comment by Czaki** [comment]
To increase discoverability, we could also set `viewer.status` text on `"Hold <Alt> key when drop the file to open plugin selection dialog"` when accepting drag event here:
https://github.com/napari/napari/blob/9ea0159ad2b690556fe56ce480886d8f0b79ffae/napari/_qt/qt_viewer.py#L1070-L1085

**Comment by DragaDoncila** [comment]
> Just late to this PR here and maybe this is not the right place to comment.

Absolutely the right place to comment, and it's never too late 😊 

> In a similar fashion built-in readers/writers could be under `File->Open`, `File->Save` and file formats supported via plugins under `File->Import` and `File->Export`. That may be slightly more preferable over `File>Open with Plugin...`

I can see how this might be clearer for those who are used to other software using `Import` as a shorthand for `open non native file`. My concern with using `Import` here would be that its introducing terminology we don't use elsewhere, *and* this method actually allows you to select builtins as your plugin if you want. I totally agree the number of `Open` options is confusing though. I would like to move forward quickly with my suggestion in the description which is unifying the entire opening mechanism under one `File->Open` entry.

File: napari/_qt/menus/file_menu.py
Comment (DragaDoncila) [reply napari/_qt/menus/file_menu.py]: Good point, will make private method above and use that. Thanks @Czaki.

**Comment by DragaDoncila** [comment]
> To increase discoverability, we could also set viewer.status text on "Hold <Alt> key when drop the file to open plugin selection dialog" when accepting drag event here:

@Czaki I tried this but we only ever see `Ready` printed. When I tried just printing anything in `dragEnterEvent` I couldn't get it to ever fire - only `_enter_canvas` fires. Any ideas?

**Comment by Czaki** [comment]
@DragaDoncila 
This what I pointed out earlier is working (at least on my machine) if you have at least one layer loaded. I totally forgot about QtWelcome that has its own drag handling and also needs to overwrite:

https://github.com/napari/napari/blob/main/napari/_qt/widgets/qt_welcome.py#L139

As there are two places where the same string needs to be applied, then it should be applied by some function method, not the global variable (translation of global variable will be done on import time, not during execution).

**Comment by DragaDoncila** [comment]
@Czaki ohh yeah the overlay. I've added a private method on the `QtViewer` now to set that message, and made sure it's translated too.

File: napari/_qt/menus/file_menu.py
Comment (Czaki) [inline napari/_qt/menus/file_menu.py:243]: I would love to see some tests that will test these functions (With mocking dialog `show`/`exec`) te be sure that the menu entry will work. Maybe by calling menu items directly.

File: napari/_qt/dialogs/qt_reader_dialog.py
Comment (Czaki) [reply napari/_qt/dialogs/qt_reader_dialog.py]: this is still not adresed witch change or explanation.

File: napari/_qt/qt_viewer.py
Comment (Czaki) [inline napari/_qt/qt_viewer.py:701]: by mock `getOpenFileNames` it is possible to write a test for this.

**Review by Czaki — COMMENTED** [review]
I am worried about the lack of tests for functions called directly from the interface. Lack of such tests may lead to a silent break of GUI and require additional work from reviewers for care about this. 

If you need, I may write such test drafts.

File: napari/_qt/dialogs/qt_reader_dialog.py
Comment (DragaDoncila) [reply napari/_qt/dialogs/qt_reader_dialog.py]: Ooops, my bad this one escaped me

File: napari/_qt/menus/_tests/test_file_menu.py
Comment (Czaki) [inline napari/_qt/menus/_tests/test_file_menu.py]:
```suggestion
        if qtpy.QT5
```

**Comment by liu-ziyang** [comment]
Hi all! just trying to understand the next step here. It seems the comments are resolved and the PR is ready to be merged. Are we waiting on some final review from Juan or?

**Comment by Czaki** [comment]
I should have access to my computer tommorow so I should do final revision then.

**Review by alisterburt — COMMENTED** [review]
Looking great here @DragaDoncila and @Czaki - @Czaki I'm pinging you here for review as well as self-assigning

edit: I would like to get this in for 0.4.17

File: napari/_qt/menus/_tests/test_file_menu.py
Comment (Czaki) [inline napari/_qt/menus/_tests/test_file_menu.py]:
```suggestion
    raise ValueError(f'Could not find action "{action_text}"')  # pragma: no cover
```

**Review by Czaki — APPROVED** [review]

**Comment by liu-ziyang** [comment]
this seems ready for merge then! going to leave it for 24h and will merge if no one else does before I do

**Comment by alisterburt** [comment]
Re-ran a failed test which looked spurious

**Pull request merged**
